apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java

```diff
@@ -32,209 +65,403 @@
+   * @return the {@link Set} of jobs registered via {@link #registerJob(PipelineJob)} that have
+   *     {@link PipelineJob#getSchedule()} values
+   */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public Set<PipelineJob<NullPipelineJobArguments>> getScheduledJobs() {
+    Set scheduledJobs =
+        jobRegistry.values().stream()
+            .filter(j -> j.getSchedule().isPresent())
+            .collect(Collectors.toSet());
+    return scheduledJobs;
   }
 
   /**
-   * Stops this monitor: it will wait for the current polling/processing operation (if any) to
-   * complete normally, then stop any future polling/processing runs from starting. This method will
-   * block until any current operations have completed and future ones have been cancelled.
+   * Enqueues an execution of the specified {@link PipelineJob}, with the specified parameters.
    *
-   * <p><strong>Note:</strong> This might block for a while! Some data sets have terabytes of data,
-   * and there is no way to safely stop processing in the middle of a data set.
+   * @param jobRecord the {@link PipelineJobRecord} of the job to run
+   * @return <code>true</code> if the specified job was enqueued, or <code>false</code> if it could
+   *     not be (e.g. because {@link #stop()} has been called)
    */
-  public void stop() {
-    LOGGER.debug("Stopping...");
+  public <A extends PipelineJobArguments> boolean enqueueJob(PipelineJobRecord<A> jobRecord) {
+    // First, find the specified job.
+    @SuppressWarnings("unchecked")
+    PipelineJob<A> job = (PipelineJob<A>) jobRegistry.get(jobRecord.getJobType());
+    if (job == null)
+      throw new IllegalArgumentException(
+          String.format("Unknown or unregistered job type '%s'.", jobRecord.getJobType()));
 
-    // If we haven't started yet, this is easy.
-    if (ccwRifLoadJobFuture == null) {
-      return;
+    /*
+     * Design Note: Java's Executor framework is great at running things and returning results, but
+     * provides no built-in facilities at all for monitoring and tracking the things that have been
+     * run. We need that monitoring here, so we use PipelineJobWrapper to catch these events:
+     * enqueue, start, complete-successfully, complete-with-exception. (See Futures.addCallback(...)
+     * a little bit further below for a discussion of some additional monitoring we need and add
+     * in.)
+     */
+    PipelineJobWrapper<A> jobWrapper = new PipelineJobWrapper<>(job, jobRecord);
+
+    // Ensure code below doesn't accidentally use the unwrapped job.
+    job = jobWrapper;
+
+    // Submit the job to be run!
+    ListenableFuture<PipelineJobOutcome> jobFuture;
+    try {
+      jobFuture = jobExecutor.submit(jobWrapper);
+    } catch (RejectedExecutionException e) {
+      // Indicates that the executor has been shutdown.
+      return false;
     }
+    jobActiveHandles.put(jobRecord.getId(), new PipelineJobHandle<>(jobWrapper, jobFuture));
+    jobsRunningCounter.incrementAndGet();
```

Comment:
Interesting thought: Does it make sense to even have a separate `AtomicInteger` here?  Since you are using a `ConcurrentHashMap` from reading `size()` it seems to have the same guarentees as `AtomicInteger` does and might actually provide more accuracy.

For example in this case, say you perform the `put` and another thread reads `jobRunningCounter`.  That will result in `jobsRunningCounter` being less than the number running but `jobActiveHandles.size()` will be correct.  

Now you have a written contract above in documentation that says this can only be incremented or read from one thread, so this works .. but it might still be simpler to just use `.size()`

Reply:
That's potentially a good catch. I added `jobActiveHandles` very late in development and didn't consider whether or not it could replace `jobsRunningCounter`. I'll ponder that a bit.

Reply:
Yup: it was safe to do so! And in looking through things to make that change, I caught a race condition in `PipelineManager.stop()` that I'd missed, so thanks!